File: projects/examples/grpc/example_grpc_client.cc
/**
 * Example gRPC client
 * This program connects to the example gRPC server and periodically calls a method indefinitely
 * References:
 * - https://grpc.io/docs/languages/cpp/basics/#client
 * - https://github.com/grpc/grpc/tree/master/examples/cpp/error_handling
 * - https://grpc.io/blog/deadlines/#c
 */

#include <chrono>
#include <iostream>
#include <memory>
#include <string>
#include <thread>

#include "absl/flags/flag.h"
#include "absl/flags/parse.h"
#include "grpcpp/grpcpp.h"
#include "projects/examples/grpc/proto/example.grpc.pb.h"

/**
 * Responsible for acting as an example gRPC client that simply sends a request containing any message on demand
 * to the example request service
 */
class ExampleGrpcClient {
 public:
  /**
   * @param[in] addr: Server address to connect to on a method call
   * @param[in] port: Server port to connect to on a method call
   * @param[in] timeout: Transaction timeout on a method call
   */
  ExampleGrpcClient(const std::string& addr, uint16_t port, std::chrono::milliseconds timeout = std::chrono::milliseconds(5000)) : timeout_(timeout) {
    const std::string target = addr + ":" + std::to_string(port);
    const std::shared_ptr<grpc::Channel> channel = grpc::CreateChannel(target, grpc::InsecureChannelCredentials());
    request_service_stub_ = com::example::grpc::RequestService::NewStub(channel);
  }

  /**
   * Transact a request for a response
   * @param[in] message: Any human readable message
   * @return true if the message was able to connect and transact a request/response
   */
  bool send_request(const std::string& message) {
    bool sent = false;

    grpc::ClientContext context;
    com::example::grpc::Request request;
    com::example::grpc::Response response;

    context.set_deadline(std::chrono::system_clock::now() + timeout_);

    request.set_msg(message);
    grpc::Status status = request_service_stub_->Send(&context, request, &response);

    if (status.ok()) {
      sent = true;
      std::cout << "Received response: \"" << response.echo_msg() << "\" (Status Code: " << response.status_code() << ")" << std::endl;
    } else {
      std::cout << "Failed to send request after timeout (" << timeout_.count() << " ms)" << " Error: " << status.error_message() << std::endl;
    }

    return sent;
  }

 private:
  std::chrono::milliseconds timeout_;
  std::unique_ptr<com::example::grpc::RequestService::Stub> request_service_stub_;
};

ABSL_FLAG(std::string, addr, "localhost", "Connect address of gRPC server");
ABSL_FLAG(uint16_t, port, 30000, "Connect port of gRPC server");

int main(int argc, char* argv[]) {
  int exit_code = EXIT_FAILURE;

  absl::ParseCommandLine(argc, argv);

  ExampleGrpcClient client(absl::GetFlag(FLAGS_addr), absl::GetFlag(FLAGS_port));

  uint8_t message_counter = 0U;
  while (true) {
    const std::string message = "Test message (counter: " + std::to_string(message_counter) + ")";
    ++message_counter;
    (void)client.send_request(message);
    std::this_thread::sleep_for(std::chrono::seconds(3));
  }

  exit_code = EXIT_SUCCESS;

  return exit_code;
}

File: projects/examples/grpc/example_grpc_client.py
#!/usr/bin/env python3
# References:
# - https://grpc.io/docs/languages/python/basics/#client
import sys
from argparse import ArgumentParser, RawTextHelpFormatter
from pathlib import Path
import time

import grpc
import projects.examples.grpc.proto.example_pb2 as example_pb2
import projects.examples.grpc.proto.example_pb2_grpc as example_pb2_grpc

DESCRIPTION = """\
Example gRPC client
This program connects to the example gRPC server and periodically calls a method indefinitely

Example usage:
./{filename}
""".format(
    filename=Path(__file__).name
)


class ExampleGrpcClient:
    """
    Responsible for acting as an example gRPC client that simply sends a request containing any message on demand to the example request service
    """

    def __init__(self, addr: str, port: int, timeout: float = 5.0):
        """
        :param addr: Server address to connect to on a method call
        :param port: Server port to connect to on a method call
        :param timeout: Transaction timeout on a method call in seconds
        """
        self._addr = addr
        self._port = port
        self._timeout = timeout

        channel: grpc.Channel = grpc.insecure_channel(f"{addr}:{port}")
        self._request_service_stub = example_pb2_grpc.RequestServiceStub(channel)

    def send_request(self, message: str) -> bool:
        """
        Transact a request
        :param message: Any human readable message
        :return: True if the message was able to connect and transact a request/response
        """
        success = False
        request = example_pb2.Request(msg=message)

        try:
            response: example_pb2.Response = self._request_service_stub.Send(
                request,
                timeout=self._timeout,
            )
            print(f"Received response:\n{response}")
            success = True
        except grpc.RpcError as err:
            print(
                f"Unable to send to {self._addr}:{self._port} within {self._timeout} s. Error: {err}"
            )

        return success

    @property
    def timeout(self) -> float:
        return self._timeout

    @timeout.setter
    def timeout(self, new_timeout):
        self._timeout = new_timeout


def get_args():
    argument_parser = ArgumentParser(
        description=DESCRIPTION, formatter_class=RawTextHelpFormatter
    )
    argument_parser.add_argument(
        "-a",
        "--addr",
        default="localhost",
        help="Connect address of gRPC server",
    )
    argument_parser.add_argument(
        "-p",
        "--port",
        type=int,
        default=30000,
        help="Connect port of gRPC server",
    )
    return argument_parser.parse_args()


def main():
    exit_code = 1
    args = get_args()

    gprc_client = ExampleGrpcClient(
        addr=args.addr,
        port=args.port,
    )

    try:
        counter = 0
        while True:
            message = f"Test message (counter: {counter})"
            counter += 1
            gprc_client.send_request(message)
            time.sleep(3)
    except KeyboardInterrupt:
        exit_code = 0

    return exit_code


if __name__ == "__main__":
    sys.exit(main())
